LibZipSharp/Xamarin.Tools.Zip/ZipArchive.cs

```diff
@@ -773,34 +788,61 @@ internal static unsafe Int64 stream_callback (IntPtr state, IntPtr data, UInt64
 					return (Int64)sizeof (Native.zip_stat_t);
 
 				case SourceCommand.Tell:
-				case SourceCommand.TellWrite:
 					return (Int64)stream.Position;
+				case SourceCommand.TellWrite:
+					return (Int64)destination.Position;
 
 				case SourceCommand.Write:
 					buffer = ArrayPool<byte>.Shared.Rent (length);
 					try {
 						Marshal.Copy (data, buffer, 0, length);
-						stream.Write (buffer, 0, length);
+						destination.Write (buffer, 0, length);
 						return length;
 					} finally {
 						ArrayPool<byte>.Shared.Return (buffer);
 					}
 
 				case SourceCommand.SeekWrite:
-				case SourceCommand.Seek:
 					Native.zip_error_t error;
-					UInt64 offset = Native.zip_source_seek_compute_offset ((UInt64)stream.Position, (UInt64)stream.Length, data, len, out error);
-					stream.Seek ((long)offset, SeekOrigin.Begin);
+					Int64 offset = Native.zip_source_seek_compute_offset ((UInt64)destination.Position, (UInt64)destination.Length, data, len, out error);
+					if (offset < 0) {
+						return offset;
+					}
+					if (offset != stream.Seek (offset, SeekOrigin.Begin)) {
+						return -1;
+					}
+					break;
+				case SourceCommand.Seek:
+					offset = Native.zip_source_seek_compute_offset ((UInt64)stream.Position, (UInt64)stream.Length, data, len, out error);
+					if (offset < 0) {
+						return offset;
+					}
+					if (offset != stream.Seek (offset, SeekOrigin.Begin)) {
+						return -1;
+					}
 					break;
 
 				case SourceCommand.CommitWrite:
-					stream.Flush ();
+					destination.Flush ();
+					stream.Position = 0;
+					destination.Position = 0;
+					stream.SetLength (destination.Length);
+					destination.CopyTo (stream);
+					stream.Position = 0;
+					destination.Dispose ();
```

Comment:
I would add a `stream.Flush ()` here, just to be sure the data is there (maybe `CopyTo` does that, maybe not, it's better to be safe than sorry)